<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Exercise 4: Graphing Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Tutorial: Basic R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tutorial Overview</a>
</li>
<li>
  <a href="exercise_1.html">Exercise 1: Getting Started</a>
</li>
<li>
  <a href="exercise_2.html">Exercise 2: Data-Structures</a>
</li>
<li>
  <a href="exercise_3.html">Exercise 3: Writing Scripts</a>
</li>
<li>
  <a href="exercise_4.html">Exercise 4: Graphing Data</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Exercise 4: Graphing Data</h1>

</div>


<p><br></p>
<p>R is great for making graphics of your data. It has a base function <code>plot()</code> that can make all sorts of graphs, but an Iowa State Alumn from the Statistics department, Hadley Wickham, created a packaged <code>ggplot</code> using the grammar of graphics to enhance the functionality of making plots in R. The package is onto its second version <code>ggplot2</code>. This tutorial will be working with <code>ggplot2</code>, so we will need to install and load the package if we don’t have it already.</p>
<pre class="r"><code>library(ggplot2)</code></pre>
<pre><code>## RStudio Community is a great place to get help:
## https://community.rstudio.com/c/tidyverse.</code></pre>
<p>And let’s make sure that we have our practice data loaded.</p>
<pre class="r"><code>readRDS(file = &quot;merged_data.RDS&quot;)</code></pre>
<div id="plotting" class="section level2">
<h2>Plotting</h2>
<p>The call to make a plot with ggplot2 is.. <code>ggplot()</code> so let’s try it.</p>
<pre class="r"><code>ggplot(merged_data)</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-5-1.png" width="576" /></p>
<p>ggplot has aesthetic arguments <code>aes()</code> that tell it what the plot should look like. Two of the arguments for the <code>aes()</code> are what the data should be for the axes <code>x</code> and <code>y</code>.</p>
<p>A plot called with ggplot can be called all at once but it’s good practice so assign the call to a variable and build on it from there. We will do this in this tutorial and hopefully it will be clear why this is a useful practice. The most common variable you will see people assign a plot to os <code>p</code>.</p>
<pre class="r"><code>p &lt;- ggplot(merged_data, aes(x = Fire_history, y = Normalized.abundance))</code></pre>
<p>**Note that ggplot knows that <code>Fire_history</code> is a reference to a data column in <code>merged_data</code> since <code>merged_data</code> was declared as the dataset in the argument.</p>
<p>and then we can call the plot with just the variable</p>
<pre class="r"><code>p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-7-1.png" width="576" /></p>
<p>All that does is make our plotting area. The way that ggplot works is that you start with your plot and then you add layers to is, so we will build on this base layer called with <code>ggplot()</code>.</p>
</div>
<div id="geometric-layers" class="section level2">
<h2>Geometric Layers</h2>
<p>The geometric layers refers to the type of plot that we want. There are over 30 types of layers, but some of the most common ones to use are</p>
<p><code>geom_point()</code> <code>geom_line()</code> <code>geom_bar()</code> <code>geom_boxplot()</code></p>
<p>So let’s try one with ours. We add layers to our ggplot using <code>+</code> then the function call. ### test</p>
<pre class="r"><code>p &lt;- p + geom_point()
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-8-1.png" width="576" /></p>
<p>So now we have some data points on our graph! But what if we decided we wanted it to be a boxplot instead?</p>
<pre class="r"><code>p &lt;- p + geom_boxplot()
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-9-1.png" width="576" /></p>
<p>Looks good! Except the points are still there.. you can add multiple geom layers to a graph so now we have both the points and the boxes. Once a geom is called we need to start the plot over if we decide we do not like it. So let’s reset our plot.</p>
<pre class="r"><code>p &lt;- ggplot(merged_data, aes(x = Fire_history, y = Normalized.abundance))</code></pre>
<p>Your geom can take arguments to change things about it also, such as color and fill.</p>
<pre class="r"><code>p &lt;- p + geom_boxplot(color = &quot;blue&quot;)
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-11-1.png" width="576" /></p>
<pre class="r"><code>p &lt;- p + geom_boxplot(fill = &quot;green&quot;)
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-11-2.png" width="576" /></p>
<p>Notice the difference in what these two options do.. and also notice that since these are layers.. when we call <code>geom_boxplot()</code> again it covers the first one, so all the arguments we want need to be in the first call.</p>
<pre class="r"><code>p &lt;- ggplot(merged_data, aes(x = Fire_history, y = Normalized.abundance))
p + geom_boxplot(color = &quot;blue&quot;, fill = &quot;green&quot;, outlier.color = &quot;red&quot;)</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-12-1.png" width="576" /></p>
</div>
<div id="colors-in-r" class="section level2">
<h2>Colors in R</h2>
<p>At its base, R uses the hexadecimal system for calling colors, but some colors have default values as we saw with <code>&quot;blue&quot;</code> and <code>&quot;green&quot;</code>. These presets are stored in base-R in the <code>colors()</code> function. if you call that function it will return all 657 preset color names.</p>
<pre class="r"><code>which(colors() == &quot;blue&quot;)</code></pre>
<pre><code>## [1] 26</code></pre>
<p>Blue is the 26th color in the presets, so we can also call it like that. There is a <a href="http://research.stowers.org/mcm/efg/R/Color/Chart/">reference</a> for what these colors look like if you ever want to get creative with your color use.</p>
<p>You can also look up the <a href="https://htmlcolorcodes.com/">hexidecimal</a> values for the colors.</p>
<pre class="r"><code>p + geom_boxplot(color = colors()[26], fill = &quot;green&quot;, outlier.color = &quot;#FF3333&quot;)</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-14-1.png" width="576" /></p>
<p>There is also a package called <code>RColorBrewer</code> which is handy for creating colorsets to use for your graphics.</p>
</div>
<div id="aesthetics-and-colors" class="section level2">
<h2>Aesthetics and Colors</h2>
<p>There are lot of different aesthetic options to use to help show the data represented in your plot.</p>
<p>If we use the <code>geom_point()</code> layer we can show each individual data point, but if you recall from the plot above when we first did this, some of the points overlapped a lot, so to help make it easier to see each of them there is an alternate <code>geom_jitter()</code> that will spread them out across the x-axis.</p>
<pre class="r"><code>p + geom_jitter(size = 2, width = 0.2, aes(color = Temperature))</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-15-1.png" width="576" /></p>
<p>When you assign <code>color =</code> in <code>aes()</code> it is assigning the categories for the range of colors, based on preset color palletes in R, so here we said we want a new color for the range of values in <code>Temperature</code>.</p>
<p>We can do it by categories of the data we are plotting too.</p>
<pre class="r"><code>ggplot(merged_data, aes(x = Fire_history, y = Normalized.abundance, fill = Fire_history)) + 
  geom_boxplot()</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-16-1.png" width="576" /></p>
<p>To change these colors we can do this</p>
<pre class="r"><code>p + geom_jitter(size = 2, width = 0.2, aes(color = Temperature)) +
  scale_color_continuous(low = &quot;yellow&quot;, high = &quot;red&quot;)</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-17-1.png" width="576" /></p>
<p>The most informative graph we can make so far combines everything we have seen so far (always be careful not to make a figure “too busy” however, it is possible to display and overwhelming amount of information)</p>
<pre class="r"><code>p &lt;- ggplot(merged_data, aes(x = Fire_history, y = Normalized.abundance))
p &lt;- p + geom_boxplot(color = &quot;black&quot;, fill = &quot;grey&quot;, outlier.shape = NA )
p &lt;- p + geom_jitter(size = 1.5, width = 0.2, aes(color = Temperature))
p &lt;- p + scale_color_continuous(low = &quot;yellow&quot;, high = &quot;red&quot;)
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-18-1.png" width="576" /></p>
<p>**Note that boxplot needs to be called first so that it is not layered over the points.. try it in the other order to see what I mean.</p>
</div>
<div id="facets" class="section level2">
<h2>Facets</h2>
<p>What if we want to display which point belongs to each gene in our dataset? We have run out of thing to color, we could use shapes but we have 9 genes and that could be “too busy”.. mkaing 9 plots individuall is cumbersome. Facets allow for multiple panels in your plot.</p>
<pre class="r"><code>p + facet_wrap(~ Gene)</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-19-1.png" width="576" /></p>
<p><code>~</code> is used to mean ‘modeled by’.</p>
<p>There are other options for facest you can explore, but the <code>facet_wrap()</code> is extremely useful.</p>
</div>
<div id="labels" class="section level2">
<h2>Labels</h2>
<p>Labels on figures are the #1 thing that people will critique.. because they are and should be the easiest thing to understand. Therefor it is important to make your labels look nice. Sometimes this is just easier in post-production using some sort of graphic editor.. but we have lots of options for it in ggplot.</p>
<pre class="r"><code>p &lt;- p +   ggtitle(&quot;Centralia Coal Mine Fires&quot;) +
  ylab(&quot;Normalized Abundance&quot;) +
  xlab(&quot;Fire Status&quot;) +
  labs(color = &quot;Temperature (°C)&quot;) 
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-20-1.png" width="576" /></p>
</div>
<div id="themes" class="section level2">
<h2>Themes</h2>
<p>There are 10 <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">preset themes</a> in ggplot. Selecting one sets a lot of options, but you can also manually set these yourself. Alternatively you can pick a theme and then tweak each option as you want from there. The most commonly used theme is <code>theme_bw()</code></p>
<pre class="r"><code>p &lt;- p + theme_bw()
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-21-1.png" width="576" /></p>
<p>Very subtle differences from what we had.. but we can change things from here. For whatever reason, by default the plot titles in ggplot are left-aligned. I persoinally like my titles centered. And I feel like our x-axis label is redundant information, so I want to remove that. Sometimes we don’t want the legend either. This can all be done in the theme options.</p>
<pre class="r"><code>p + theme(plot.title = element_text(hjust = 0.5),
      axis.title.x = element_blank(),
      legend.position=&quot;none&quot;
      )</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-22-1.png" width="576" /></p>
<p>Theme is where all the subtle changed occur, and is very important. There are lots and lots of options in theme and worth the time to explore when you want to publish a figure. Almost always.. if you want to do it, there is a way.</p>
</div>
<div id="bar-charts" class="section level2">
<h2>Bar Charts</h2>
<p>Bar charts are also very popular for displaying data, particularly abundances like we were working with before. Bar charts require at least one axis to be defined with a dataset, then the other will by default be counts of datapoints in that set.</p>
<pre class="r"><code>ggplot(merged_data, aes(x = Fire_history)) + 
  geom_bar()</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-23-1.png" width="576" /></p>
<p>We can color each bar in the char like we saw before.</p>
<pre class="r"><code>ggplot(merged_data, aes(x = Fire_history, fill=Fire_history)) + 
  geom_bar()</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-24-1.png" width="576" /></p>
<p>But this is redundant information. We can’t color it by a <code>y</code> variable because our <code>y</code> is counts. So lets define our <code>y</code> as something more meaningful. To do this we need to change the <code>stat =</code> argument in our <code>geom_bar()</code>.</p>
<pre class="r"><code>ggplot(merged_data, aes(x = Fire_history, y = Normalized.abundance)) + 
  geom_bar(stat = &quot;identity&quot;)</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-25-1.png" width="576" /></p>
<p>Having 3 bars is not really maximizing the usefulness of our bar chart, so lets graph a different metric.</p>
<pre class="r"><code>ggplot(merged_data, aes(x = Site, y = Normalized.abundance)) + 
  geom_bar(stat = &quot;identity&quot;, color = &quot;black&quot;, aes(fill = Gene)) </code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-26-1.png" width="576" /></p>
<p>Sometimes stacking the <code>y</code> doesn’t make much sense.. so we can “dodge” the bars. Also you may have noticed that our x is harder to read now becasue there are more labels, so we can fix that with <code>theme()</code></p>
<pre class="r"><code>p &lt;- ggplot(merged_data, aes(x = Site, y = Normalized.abundance)) + 
  geom_bar(position = &quot;dodge&quot;, stat = &quot;identity&quot;, color = &quot;black&quot;, aes(fill = Gene)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
p</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-27-1.png" width="576" /></p>
<p>But there needs to be a balance between information presentation and becoming too much.</p>
</div>
<div id="saving-figures" class="section level2">
<h2>Saving Figures</h2>
<p>Now that we have spent all this time creating the perfect graph we need to figure out how to do, arguably, the most important part.. getting it out of R.</p>
<p>The way to do this in base R is easy once you are used to it, but fairly unintuitive. Firstly you need to know about different image types</p>
<p><code>.png</code> - Portable Network Graphs</p>
<p><code>.tiff</code> - Tagged Image File Format</p>
<p><code>.jpeg</code> - Joint Photographic Experts Group</p>
<p><code>.raw</code> - raw</p>
<p>It’s a lot to know the attributes for all of these files. There are varrying opinions on which to use.. I recommend <code>.png</code> for its balance of detail and file size.</p>
<p>to save a file in R you use any of the functions with these file extensions name. Within the function you need the path/filename of where you want to save the image. The function has default settings for image size and resolution. If you are making a figure for a paper or poster then you need a large image with high resolution. You can adjust arguments for this. I am comfortable with inches, so I set <code>units = &quot;in&quot;</code>, but there is also pixels “px”, centimeters “cm”, and milimeters “mm”. For high-resolution I recommend setting <code>res = 300</code></p>
<pre class="r"><code>png(filename = &quot;My_Graph.png&quot;, width = 9, height = 5, units = &quot;in&quot;, res = 300)</code></pre>
<p>Calling this function doesn’t save the graph though. It opens what R calls a ‘device’. This means that the next time you call a graphic in R it will open that graphic to that device (by default the null device is the console). Once we call our graph we then need to use <code>dev.of()</code> to close the png device and return to our console. If we do not, then R will continue saving the next graphics we call in this session to the file we specified.</p>
<pre class="r"><code>p
dev.off()</code></pre>
<p>Now we should have our graph!</p>
<p>Another way to save this is with a function in ggplot.. which is for sure friendlier.</p>
<p><code>ggsave()</code> will automatically save the last plot you called before this by default (can be changed with argument <code>plot =</code>. It automatically opens and closes the device, and it defaults to a <code>res = 300</code>. It can take the same <code>height=</code>, <code>wdith=</code>, and <code>units=</code> arguments also. Additionally it will automatically use the device that matches the file extension in your <code>file =</code>.</p>
<pre class="r"><code>p
ggsave(filename = &quot;My_Graph.png&quot;, height = 5, width = 5, units = &quot;in&quot;)</code></pre>
</div>
<div id="final-practice" class="section level2">
<h2>Final Practice</h2>
<p>To sum it all up, try reproducing this graph from the dataset <code>final_practice_problem.RDS</code>. This is data from a project that I worked on and it should look like this.</p>
<pre><code>##      Sample      OTU Abundance Treatment Month
## 2185     X5    OTU_2     53767    Bagged   May
## 3060   X652    OTU_2     29150    Bagged  June
## 2193     X5    OTU_5     19977    Bagged   May
## 4705    X87 OTU_4036     15324   Exposed  June
## 3068   X652    OTU_5     13588    Bagged  June
## 2010   X473    OTU_2     13222    Bagged  June</code></pre>
<p><img src="exercise_4_files/figure-html/unnamed-chunk-32-1.png" width="576" /></p>
<p><br><br><br><br> Good luck in your data analysis! And if you can’t find the answer online, never hesitate to ask a friend.</p>
</div>


<p><br>
<br>
<strong><a href="https://schuyler-smith.github.io/" target="_blank">Schuyler Smith</a></strong>
<br>
Ph.D. Student - Bioinformatics and Computational Biology<br>
Iowa State University.  Ames, IA.</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
